DESIGN.md

```diff
@@ -0,0 +1,362 @@
+potential return values. It returns a list because symbolic evaluation may
+branch and explore different paths.
+
+#### The key function — `symEvalOneStep`
+
+[`symEvalOneStep`]: #the-key-function-symEvalOneStep
+
+The key function in this module is `symEvalOneStep`, whose type is the
+following:
+
+```haskell
+symEvalOneStep ::
+  forall lang .
+  SymEvalConstr lang =>
+  TermMeta lang SymVar ->
+  WriterT Any (SymEval lang) (TermMeta lang SymVar)
+```
+
+Once again, morally speaking, forgetting the `lang` type parameters and
+replacing `TermMeta lang SymVar` by just `Term`, this function has type `Term ->
+SymEval Term`. One can then think of this as a function of type:
+
+```haskell
+SymEvalEnv -> (Term, SymEvalSt) -> [(Term, SymEvalSt)]
+```
+
+Such a function takes a term in the input language and a symbolic evaluation
+state. It tries to apply one step of beta-reduction to the term, which may yield
+zero, one or several potential terms and updated states.
+
+As an example, let us assume that our input term is the body of the `foldr`
+function in [the example language], specialised for lists of `Integer`s (for
+readability, we remove the type applications).
+
+```haskell
+match_List l
+  e
+  (\(x : Integer) (xs : List Integer) . f x (foldr f e xs))
+```
+
+and that our state specifies that:
+
+- `e` is an integer, without giving its value,
+- `f` is the function `\(x : Integer) (y : Integer) . x + y`, and
+- `l` is a list of integers, without giving its value.
+
+One step of the symbolic engine would then have to explore both branches of the
+pattern matching, and therefore it would return two pairs `Term, SymEvalSt`:
+
+1. In the first pair, the `Term` would be `e` itself and the state would be
+   updated to account for the fact that we now know that `l = Nil`.
+2. In the second pair, the `Term` would be `f x (foldr f e xs)` and the state
+   would be updated to account for the fact the we now have two more variables,
+   `x` and `xs`, that are an integer and a list of integers respectively, and an
+   additional constraint, `l = Cons x xs`.
+
+The whole symbolic execution consists in running this one step over and over
+again until it is not possible to reduce the terms anymore. In the example
+above, our first term/state is now blocked because there isn't much one can do
+with an integer. The second term/state is not blocked because one can replace
+`f` by its definition and continue unfolding from there.
+
+#### The symbolic evaluation state — `SymEvalSt`
+
+Defined in [`Pirouette.Symbolic.Eval.Types`], the symbolic state basically
+contains the following:
+
+1. A context of variables, that is a map from variable names to their types.
+   This context is typically called Γ in literature, and `gamma` in the code.
+
+2. A set of constraints linking variables together, for instance `l == Cons x
+   xs`, `k != Nil` or `x = 7`. It is in fact not a set, but a complex
+   union-find-based type.
```

Comment:
```suggestion
   xs`, `k != Nil` or `x = 7`. It is implemented with a union-find-based type.
```
A reference about `union-find` could be helpful here.

Reply:
Do you have an idea of a good reference for this?

In term of academic references, apparently, the first reference could be [(Galler & Fisher 64)](https://dl.acm.org/doi/10.1145/364099.364331) but I don't think I've ever read it. I have read stuff by Ullman, probably in a book of his, which might make it more readable. I'll try to find that. I have also read [(Tarjan & van Leeuwen 84)](https://dl.acm.org/doi/10.1145/62.2160) and that's I think the implementation that is used nowadays; it's been proven later that it was optimal.

But maybe you just meant a reference to an overview of Union-Find data structures? A wikipedia page, for instance?

Reply:
> But maybe you just meant a reference to an overview of Union-Find data structures? A wikipedia page, for instance?

This might do.

Reply:
Done in d7c953e.